### ISSUE TITLE
Server-Side Rendering (SSR) with Next.js

### ISSUE THREAD

**Issue by ericclemmons** [aws-amplify/amplify-js#5435]
Given Next.js’ popularity, customer impact, and being a hybrid framework supporting both SSR & SSG, solving for Next.js first unblocks other frameworks (e.g. Gatsby, Nuxt).

There is additional work to be done for CLI & Console support (particularly around deployments), but this Epic focuses on JS support.

**This feature is live! See the [Getting Started](https://docs.amplify.aws/start/q/integration/next) tutorial and [SSR docs](https://docs.amplify.aws/lib/ssr/q/platform/js) for the latest!**

## Public Preview

We'd love your help testing out Amplify's `API`, `Auth`, and `DataStore` categories in your Next.js app!

1. First, reinstall Amplify using the `@preview` tag on NPM:

	```shell
	yarn add aws-amplify@preview
	# or
    npm install aws-amplify@preview --save
	```

1. If you're only accessing public data (e.g. `API.graphql(listBlogs)`), you're done! 😌 

1. For secure access the current user on the server, you'll need to scope Amplify to the current request (and _only_ that request) using our new `withSSRContext` helper:

	```diff
	diff --git a/my-next-app/pages/index.tsx b/my-next-app/pages/index.tsx
	index 22b0c471..d7ce454b 100644
	--- a/my-next-app/pages/index.tsx
	+++ b/my-next-app/pages/index.tsx
	@@ -1,6 +1,6 @@
	 import { CognitoUser } from '@aws-amplify/auth'
	 import { GRAPHQL_AUTH_MODE } from '@aws-amplify/api-graphql'
	-import { Amplify, API, Auth } from 'aws-amplify'
	+import { Amplify, withSSRContext } from 'aws-amplify'
	 import { Authenticator } from 'aws-amplify-react'
	 import { GetServerSideProps } from 'next'
	 import Head from 'next/head'
	@@ -13,8 +13,6 @@ import { CreateTodoMutationVariables } from '../src/API'
	
	 Amplify.configure(awsconfig)
	
 	 // 👇 Your frontend code is now "SSR-aware", giving the backend access to the current session
	+const { Auth, API } = withSSRContext()
	+
	 export default function Home(props) {
	@@ -392,8 +390,6 @@ export default function Home(props: Props) {
	 }
	
	 export const getServerSideProps: GetServerSideProps = async (context) => {
       // 👇 Auth & API are now scoped to the _current_ user's session
	+  const { Auth, API } = withSSRContext(context)
	+
	```

	For more information on `withSSRContext`, see https://github.com/aws-amplify/amplify-js/pull/6146.

1. Last step!  **Let us know your experience in the comments below** 👇 

## Milestones

- [x] SSR `withAuthenticator` (#5138)

	`aws-amplify-react` and `@aws-amplify/ui-react` should at least render `withAuthenticator` and `<AmplifyAuthenticator />` on the server. This was previously blocked by an `import "...css"`, but with Amplify v3 this was moved to customer apps to import.

	This doesn't mean session state yet, but unblocking the SSR process.

- [x] SSG with `API`

	Support API calls (e.g. `GraphQLAPI.graphql(...)`) in `getStaticProps`.

- [x] SSR with `API`

	Support API calls (e.g. `GraphQLAPI.graphql(...)`) in `getServerSideProps`.

- [x] Next.js `/pages/api/*` routes with API

	Similar to SSG & SSR, these routes are lambdas that need to support `API.graphql` calls.

- [x] SSR with `Auth` (#5710)

	Support shared client & server state so that authentication flows on the client share credentials with the server (via a subset of cookies).  Calls to `Auth.currentAuthenticatedUser()` should succeed on the server the same way they do on the client.

- [x] SSR with `DataStore` (#5450) 

	`DataStore` needs special attention (due to WebSockets) to work on the server, and calls to `DataStore.query` have to wait for the sync process to complete (otherwise results are initially empty).

- ~~[ ] SSR with Analytics (https://github.com/aws-amplify/amplify-js/issues/6208)~~

- [x] `withSSRContext` (https://github.com/aws-amplify/amplify-js/pull/6146)

---	

Related milestone: https://github.com/aws-amplify/amplify-js/milestone/26
Related issues: #5101, #4178, #3741, #1613, #3278, #5121, #5097, #4972, #2230, #4990, #4851, #5293, #5435, #5322, #3854, #3053, #3348, #5138, #4311, #4305, #4207, #3037, #992

**Comment by dabit3** [comment]
@ericclemmons you are the man.

**Comment by sovattha** [comment]
It took me a day to realize that Amplify does not work well with the SSR of Next.js.

Going to the competition (Auth0) for the time being.

**Comment by dabit3** [comment]
@sovattha check back with us soon, this is already being developed and tested.

**Comment by ericclemmons** [comment]
@sovattha Do any of issues in https://github.com/aws-amplify/amplify-js/milestone/26 match your experience?  `withAuthenticator` works with the latest release:

1. Create or Update `pages/_app.tsx` with `Amplify.configure`:

   ```tsx
   import '@aws-amplify/ui/dist/style.css'
   import Amplify from 'aws-amplify'
   import { AppProps } from 'next/app'

   import awsconfig from '../src/aws-exports'

   Amplify.configure(awsconfig)

   function MyApp({ Component, pageProps }: AppProps) {
     return <Component {...pageProps} />
   }

   export default MyApp
   ```

1. Wrap `pages/index.tsx` with `withAuthenticator`:

   ```tsx
   export default withAuthenticator(Home)
   ```

If you have any issues, feel free to create a new issue so I can track it here 🙏

**Comment by sovattha** [comment]
I'll keep an eye on your progress guys. The progress of the project is exciting. Thanks for noticing about the milestone.

**Comment by alexandrzavalii** [comment]
Hi @ericclemmons , thanks for taking a look at the issue!
I still see the error in Next.js `ReferenceError: window is not defined`
I am using `"aws-amplify": "^3.0.8",`, should I be using `unstable`?

**Comment by ericclemmons** [comment]
@alexandrzavalii I saw the same thing here: https://github.com/aws-amplify/amplify-js/issues/5293#issuecomment-610673005 (But the app still works)

If your Next.js is getting a different TypeScript/compilation/breaking error, open up a new issue and tag me: @ericclemmons.  I'm happy to take a look!

**Comment by alexandrzavalii** [comment]
@ericclemmons in Chrome console I got this error as well
```
        Error: Amplify has not been configured correctly.
            This error is typically caused by one of the following scenarios:
```
No idea if it is caused by SSR or sthg else

**Comment by ericclemmons** [comment]
@alexandrzavalii Definitely open up an issue with some more about your usage (including how/where you're doing `Amplify.configure`).  I have working examples that'll make their way into the docs once we thoroughly test it, but I'd like to unblock you if I can :)

**Comment by alexandrzavalii** [comment]
Hey @ericclemmons, Looking forward for your example! It would be very helpful to have SSR example. 
Like having a navbar rendering "signin" or "profile" depending on user status.

**Comment by TimNZ** [comment]
@ericclemmons I don't want to crap on anyone (or pick on you specifically),
but I've spent many hours looking for best practice for AWS Amplify/Cognito on how to do authentication in the browser and pass user identifying token to server i.e. the most standard of JWT type stuff for standard REST calls that don't presume API Gateway + IAM auth on calls.
Also people have been asking for guidance/support for SSR for years.

Surely there is a single doc/video on this somewhere?

I've finally sorted out extracting JWT version of Access token to pass in Authorization header, for decoding by Lambda functions, but finding out how to do all this was more painful than any other framework I've used in years.

Since Amplify by default doesn't store any session data in cookies, SSR doesn't have anything to work with.
Even finding out this was pain.

Every other framework has this most basic of scenarios working pretty much out the box and well documented.

withAuthenticator doesn't address any of this, I'm unclear as to why you reference it above.

Appreciate your thoughts on this.
I have read a lot of comments around the place that people have given up on Amplify because it's so hard.

**Comment by ericclemmons** [comment]
@TimNZ I hear you. I really do. When I joined Amplify in October, I was equally surprised that SSR wasn't addressed.

It sounds like you went through the same pain-points I did on Friday (lack of cookie support), but we're very close to having this resolved.

Regarding `withAuthenticator`, there are varying degrees of SSR support: `withAuthenticator` successfully renders out of the box (whereas it used to not!).  Your case is more complex, since it relies on sharing session state between the client & the server so that the server can make privileged calls.

I'd love to know more about your use-cases so that they're not overlooked!  Can you share more about this scenario (especially if you have sample code)?
> how to do authentication in the browser and pass user identifying token to server i.e. the most standard of JWT type stuff for standard REST calls that don't presume API Gateway + IAM auth on calls.

**Comment by TimNZ** [comment]
@ericclemmons thanks for quick response.

The use case is the classic that every framework has a flow for:

* user signs in -> verified on server -> server returns id/credentials/tokens
* client 'caches' this - cookie, localStorage etc
* clients call servers, identification data is included in calls to server

For a lot of frameworks/stacks, that's a session cookie.
That cookie might be httpOnly if it's returned by a server in a standard page flow structure i.e. not a SPA

I think the pattern I'm seeing with many people struggling is the Amplify documentation is shockingly silent on showing how this standard flow works using Amplify.
Even 3rd party guides barely address it.
Even 3rd party packages often are silent on it when they are providing a work-around for some aspect.

So many people are getting stuck on trying to get the system to work the way they are used to.
Amplify is engineered to work with a complete end to end AWS stack.

In my case (which is typical), I have an API deployed on API Gateway/Lambda, deployed via Serverless.

I 'expected' to authenticate using Amplify - doing it manually e.g. Auth.signIn etc,
and then I could get a identifying session/user token to pass to the server in API calls.

I had just presumed that the 'session token' would be in managed in a cookie so it was available in Lambda event.headers.

Having done traditional client/server dev, and SPA - Feathers JS, Sails etc,
this is what everyone is expecting, and Amplify is making real hard to understand how to replicate the flow.

Ignoring XSS concerns, now I just might extract the ID  or Access token as JWT and store it in a cookie on successful signIn.
I can now process that in my Lambda via event.headers.cookie['<my app cookie>'],
and in Next.js getInitialProps for SSR

If I wanted to apply best practice for XSS mitigation I'd implement redirect flow to endpoint that set httpOnly cookie.

^^ standard stuff

So the issue is poor documentation on how to do the above which most people are expecting kind of just works, like how they are used to with other stacks.

I think this is the root problem.

What do you think?

**Comment by crash7** [comment]
@TimNZ sorry to jump in the middle of your discussion with @ericclemmons, but I have more or less the same kind of scenarios and here are my two cents (based on my wall head banging):

a) you could create your own session and store it with an http only cookie (I have this workaround with a GraphQL server, it's not ideal but cognito handles all the auth and on successful auth, there is a login mutation that receives the jwt token, and the server verifies the token and updates de session) 

b) since you are using api gateway, you could just add an authorizer and send the access token that cognito returns in the headers, validate it in the authorizer and if everything is OK, your function will be invoked. Again, not ideal because you need to add the authorizer. 

Disclaimer: I don't use Amplify directly but I do use the aws cognito identify package that amplify uses under the hood, the main problem while doing SSRs is that we need to preload the cognito store with the data we read from the token (that gets passed in the headers) without putting a lot of if browser conditions.

**Comment by TimNZ** [comment]
@crash7 

My comments are why are so many people struggling with scenarios they just expect to work, not that they can't be solved, once you understand things.

Now I got 'expert' through the pain, I know how to address it many ways.

And the Amplify documentation doesn't help because it just mostly presumes you'll use the full AWS end-to-end flow including IAM authorizer etc.

**Comment by TimNZ** [comment]
@ericclemmons just a final follow on from my comment.

I agree with the frustration of the OP
https://github.com/aws-amplify/amplify-js/issues/3495

I'm already considering moving away from Cognito after 3 days of pain.

**Comment by odiaseo** [comment]
@TimNZ I share the same pain and sometimes shocked beyond words, I acknowledge the work the team is doing with amplify but really amplify is not production ready.

**Comment by ericclemmons** [comment]
I've updated the PR description with more information on _how_ SSR support is progressing. If you have any feedback related to SSR support & features, please share!

If you have other feedback, you can raise an issue, start a discussion, or join our Discord community:

- https://github.com/aws-amplify/amplify-js/issues/5519
- https://github.com/aws-amplify/amplify-js/issues/5716

**Comment by MontoyaAndres** [comment]
Hey @ericclemmons how can we use the `/pages/api/*` routes with `API.graphql` 😲 can we use this feature right now?

**Comment by ericclemmons** [comment]
@MontoyaAndres Yes, so long as the data is public (e.g. not requiring auth credentials, which is in progress)!  Here's a snippet from a working example of mine:

```tsx
// pages/api/blogs.tsx
import API, { graphqlOperation, GraphQLResult } from '@aws-amplify/api-graphql';
import Amplify from 'aws-amplify';

import { ListBlogsQuery } from '../../src/API';
import { listBlogs } from '../../src/graphql/queries';

import awsConfig from '../../src/aws-exports';
Amplify.configure(awsConfig);

export default async function getBlogs(req, res) {
	const { data } = (await API.graphql(
		graphqlOperation(listBlogs)
	)) as GraphQLResult<ListBlogsQuery>;

	res.send(data.listBlogs.items);
}
```

**Comment by adammanderson** [comment]
@ericclemmons, really encouraging to see progress on SSR for Auth :smile: It's been overlooked for too long!

Any ETA on when we can get to use it?

**Comment by svrcekmichal** [comment]
@ericclemmons is somewhere available example repo where we can see what part of work is done and how to setup it all?

**Comment by dtelaroli** [comment]
+1

**Comment by ericclemmons** [comment]
@adammanderson @svrcekmichal Unfortunately, not public yet! I'm addressing security concerns and don't want to risk code getting adopted without those addressed 😅

**Comment by mdegrees** [comment]
@ericclemmons is there any hack to go around this issue https://github.com/aws-amplify/amplify-js/issues/5101 while waiting for this milestone to complete. I want to use Datastore with Next js in a not SSR way for the time being.  Then go SSR when it's fully supported.

**Comment by svrcekmichal** [comment]
@ericclemmons ok, no pressure, thanks for all the time you put into this :)

**Comment by apoorvmote** [comment]
Hey guys I am interested in using nextjs with amplify(auth and api). There was an update/release yesterday to amplify-js. 
How will I know that nextjs is supported in any release/update?? 
And how will I know which features(e.g. SSG, auth, api) of Nextjs are currently supported on any release/update??

**Comment by ericclemmons** [comment]
@apoorvmote You can continue following & commenting on this issue for progress: we'll update this once it's on `latest`.

🎉 **Good news though! I just shipped a Next.js preview to `aws-amplify@preview`!**

I've updated the description at the top (https://github.com/aws-amplify/amplify-js/issues/5435) with steps to get started.  `Auth`, `API`, and `DataStore` are actively developed against, but I'm tracking other categories at the top as well.

Give it a go in your app & comment below on your experience 👇

**Comment by firrae** [comment]
This seems promising, and if done AND documented well this could be a great alternative to Auth0. We've been using Cognito at work and while I like the concept the documentation at times has been terrible, though I find that a lot across Amazon development resources, so I hope this won't meet the same blocking points of "this doesn't seem to be documented, and if it is it's not obvious".

I'm happy I stumbled on this before I started a new project going the other route. Progress seems to be going well maybe I'll bite the bullet and try out Amplify before going another route.

Also, are there currently any plans, or updates, on how things like AppSync will work with SSR? API is fine, but AppSync is what I had begun to look at if I went the all AWS route.

**Comment by apoorvmote** [comment]
I am also building static blog with Nextjs. And I will be adding my own commenting system with dynamodb and graphql. Also will be building my own custom forum with SSR. Unfortunately because of some constraints I have to connect to appsync with rest api + lambda function. 
When will amplify start supporting faster and cheaper HTTP API?

**Comment by ericclemmons** [comment]
@firrae Currently we're focusing on Amplify-specific categories (e.g. `Auth`, `API`, and `DataStore`), which will be actively tested against Next.js.

**Comment by apoorvmote** [comment]
I have tested `Auth` with `withSSRContext` and it works as advertised. But I am using `amp` pages for loging in with [amp-access](https://amp.dev/documentation/components/amp-access/?format=websites).
I need to build 3 Rest API endpoints
1. /check-auth-status
1. /login
1. /logout

I make post request on each endpoint to get login status, submit login info and logout the user. Its pretty self explanatory. But the `Auth` Stores cognito user information in local storage on browser. 

**How do I check login status at rest API with lambda?**

**Comment by ericclemmons** [comment]
@apoorvmote Is this with Next.js? (https://nextjs.org/docs/api-reference/next/amp)

**Comment by apoorvmote** [comment]
> @apoorvmote Is this with Next.js? (https://nextjs.org/docs/api-reference/next/amp)

@ericclemmons Yes. So Next.js makes it easy to implement [amp](https://amp.dev/). So the original specification/documentation is available on https://amp.dev/ . I am looking to build basic text commenting system in amp. That requires [amp-access](https://amp.dev/documentation/components/amp-access/?format=websites) component to authenticate users. 

For authentication I need to store JWT into a cookie. Because amp doesn't support easy access to local storage. Because of recent developments on [same-site cookie](https://blog.chromium.org/2019/10/developers-get-ready-for-new.html). The cookies will become more secure. 

I don't know what is the priority for you guys to support amp. So I have started learning CDK, cloudformation. So I get to implement HTTP API with my custom amp authentication. I won't be using amplify for this project.

**Comment by ericclemmons** [comment]
@apoorvmote If you're using `@preview` (https://github.com/aws-amplify/amplify-js/issues/5435#issuecomment-655660103), `/check-auth-status` should be able to return the current user via something like:

```js
import { Amplify, withSSRContext } from 'aws-amplify'
import { NextApiRequest, NextApiResponse } from 'next'

import awsconfig from 'src/aws-exports'

Amplify.configure(awsconfig)

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const { Auth } = withSSRContext({ req })

  try {
    const user = await Auth.currentAuthenticatedUser()

    return res.status(200).json({ message: "Authenticated!" })
  } catch (error) {
    console.error(error)
    return res.status(500).json({ error })
  }
}
```

`/login` & `/logout` dedicated routes aren't supported as part of this preview because authentication is delegated to Cognito on the client.

Can you open a new issue as a Feature Request with your circumstance?  It looks like `amp-access` (https://amp.dev/documentation/components/amp-access/?format=websites) would need explicit consideration as an authentication flow vs. SSR support alone via Next.js.

**Comment by apoorvmote** [comment]
@ericclemmons 
I tried something similar. Instead of building API I was testing `getServerSideProps` on login page just to verify if I am logged in on client then am I logged in on server? Something like following.
```typescript

export const getServerSideProps: GetServerSideProps = async(context) => {

  const { Auth } = withSSRContext(context)

  let user: any = null 

  console.log('get server')

  try {

    user = await Auth.currentAuthenticatedUser()

    console.log('ssr user', user)

  } catch (err) {

    console.log('error', err)
  }

  return {
    props: {
      user: JSON.parse(JSON.stringify(user))
    }
  }
}
```
I was logged in on client but on server user was null. I was able to get user on the same page on client like follows. 

```typescript
useEffect(() => {
    async function getAuth() {
      
      try {

        const user = await Auth.currentAuthenticatedUser()

        console.log(user)

      } catch (err) {

        console.log('front end error', err)
      }
    }

    getAuth()

  }, [])
```
I am new to `nextjs`. So I don't know if I have implemented any of this wrong. I refreshed page many times. But there is user on client and user is null at the server. 

>Can you open a new issue as a Feature Request with your circumstance? It looks like [amp-access](https://amp.dev/documentation/components/amp-access/?format=websites) would need explicit consideration as an authentication flow vs. SSR support alone via Next.js.

Yes I will open new issue. 

The underlying issue is we all want better SEO. So google built amp that has built in best page experience. But developers pushed against google for forcing us to use amp and not use our favorite stack. So google launched [core web vitals](https://web.dev/vitals/). So now amp is no longer the only option for better SEO. We can use our favorite stack and as long as we conform to [core web vitals](https://web.dev/vitals/)  and other best practices explained at https://web.dev/ 
And we love `amplify` because it has best practices built in so we don't have think about it. 

Now if you can deliver on that then I will give up this amp nonsense and use Nextjs directly with amplify. Until then I will build my own infrastructure with cdk and cloudformation and do server side rendering with nodejs and handlebarjs and use amp to build websites.

**Comment by ericclemmons** [comment]
@apoorvmote It sounds like you're doing all the right things!  One last check: does your client-side code also use `withSSRContext`? Otherwise, the credentials aren't made available to the server:

```js
import { Amplify, withSSRContext } from 'aws-amplify'
import awsconfig from 'src/aws-exports'

Amplify.configure(awsconfig)

const { Auth, API } = withSSRContext()

export default function Home() {
```

Otherwise, I'd need to check to see if Next.js' AMP behavior changes things somehow...

**Comment by apoorvmote** [comment]
@ericclemmons 
Does it make any difference with 
```typescript
import { Amplify, withSSRContext } from 'aws-amplify'
import config from '../src/aws-exports'

Amplify.configure(config)

export default function Homepage({ user }: iProps) {

  const { Auth } = withSSRContext()
```
vs 
```typescript
import { Amplify, withSSRContext } from 'aws-amplify'
import config from '../src/aws-exports'

Amplify.configure(config)

const { Auth } = withSSRContext()

export default function Homepage({ user }: iProps) {
  
```
Because I am using the first version.

**Comment by ericclemmons** [comment]
@apoorvmote You only need one instance of `Auth` on the client, so the first is preferred.  The second version you're using would re-create `Auth` (and all of Amplify) with each re-render of `Homepage`.

Even though it's less optimal, you're doing the right thing: so long as the authentication flow completes successfully on the client, the server should then be able to access the current user.

**Comment by apoorvmote** [comment]
@ericclemmons Thanks for clarification. Nextjs does show some promising features. But I am working on to somehow make [amp-access](https://amp.dev/documentation/components/amp-access/?format=websites) work with whatever technology that works. Then transfer the knowlege to upgrade to Nextjs with amplify. And twist Amplify to fit my workflow.

**Comment by svrcekmichal** [comment]
@ericclemmons I'm using similat setup to what @apoorvmote is using [here](https://github.com/aws-amplify/amplify-js/issues/5435#issuecomment-665389383) with no luck to get user on sever side.

What I noticed is that when I console.log Auth class in browser, there's multiple cognito cookies in UniversalStorage, but when console logged in getServerSideProps, UniversalStorage's store is almost empty, with all the cognito cookies missing. Do you maybe have some example working repo somewhere?

Posting here also my code, I've got auth route and 2 separate routes, one with SSR and one client side. On client side I've got "Hi, svrcekmichal@gmail.com" message, on server "no user":
```javascript
// /pages/user-client.tsx

Amplify.configure(awsConfig)

const { Auth } = withSSRContext()

export default function AuthPage() {
  const [loading, setLoading] = useState(true)
  const [user, setUser] = useState(null)

  useEffect(() => {

    async function getAuth() {
      try {
        setUser(await Auth.currentAuthenticatedUser())
        setLoading(false)
      } catch (e) {
        console.error(e)
        setLoading(false)
      }
    }

    getAuth()
  }, [])

  if (loading) {
    return 'Loading...'
  }

  if(!user) {
    return 'No user'
  }
  return (
    <h1>Hi, {user.attributes.email}</h1>
  )
}
```

and 

```javascript
// /pages/user-server.tsx

Amplify.configure(awsConfig)

export default function User({ user }) {
  if(!user) {
    return 'No user'
  }
  return (
    <h1>Hi, ${user.attributes.email}</h1>
  )
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const { Auth } = withSSRContext(context)
  let user = null

  try {
    user = await Auth.currentAuthenticatedUser()
  } catch (e) {
    console.error(e)
  }

  return {
    props: {
      user: JSON.parse(JSON.stringify(user))
    }
  }
}

```

**Comment by apoorvmote** [comment]
@ericclemmons 
I have bit of off topic suggestion. 
If I understand correctly you and others are working on Auth and API for NextJS with Amplify. 

If you think about it then CDK and Amplify are very similar. If you and/or others could build CDK templates for using NextJS for Auth and API as you guys work on Amplify then it would be beneficial to lot of people(mainly me  LOL). And you guys could upload them to [aws-cdk-examples](https://github.com/aws-samples/aws-cdk-examples)

I understood CDK lot faster by studying templates than reading documentation. Any partial template can save me a lot of time and headache. 

Assuming I get [amp-access](https://amp.dev/documentation/components/amp-access/?format=websites) working. I myself am planning to blog about bunch of CDK templates. Either way thanks for reading my suggestion.

**Comment by asterikx** [comment]
Is it possible to add `withSSRContext` to the individual package such as `@aws-amplify/auth`, so that you don't have to require the whole `aws-amplify` package if you only need auth for example?

**Comment by zelphir** [comment]
it looks like that server side is not working, and that is because the cookies are not saved and sent.  @svrcekmichal @apoorvmote can you confirm that the lack of cookies is causing the same issue to you guys?

**Comment by asterikx** [comment]
@ericclemmons I got that number from VSCode [Import Cost](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost) extension. I'm not sure what the final impact on the bundle is.

My other regular dependencies have ~7-15 kB uncompressed, e.g. `react` has 6.3 kB, `emotion`(CSS-in-JS) has 15.4 kB. When you look at [BundlePhobia](https://bundlephobia.com/result?p=aws-amplify@3.0.18-preview.46), the numbers are even worse.

`react@16.13.1`: [6.3 kB/2.6 kB (minified/gzipped)](https://bundlephobia.com/result?p=react@16.13.1)
`emotion@10.0.27`: [15.4 kB/5.8 kB (minified/gzipped)](https://bundlephobia.com/result?p=emotion@10.0.27)
`axios@0.19.2`: [13.1 kB/4.4 kB (minified/gzipped)](https://bundlephobia.com/result?p=emotion@10.0.27)
`aws-amplify@3.0.18-preview.46`: [852 kB/207 kB (minified/gzipped)](https://bundlephobia.com/result?p=aws-amplify@3.0.18-preview.46)
`aws-amplify@3.0.23`: [931 kB/213 kB (minified/gzipped)](https://bundlephobia.com/result?p=aws-amplify@3.0.23)

**Comment by ericclemmons** [comment]
@asterikx I would take those numbers with a grain of salt: the build output is what matters most and is where tree-shaking is visible.

Amplify v3 was built specifically with this in mind: https://twitter.com/ericclemmons/status/1240784212241846275

(Still, I am monitoring the impact of client-only `aws-amplify` vs. SSR `aws-amplify` with this PR)

**Comment by ianmartorell** [comment]
Hey @ericclemmons, thanks for the work you're putting into getting SSR working! If I understood correctly, all the improvements that are being made to make Amplify work with Next.js will still require the final app to be hosted using Vercel, since Amplify Console doesn't support SSR. Is that right? Or how would this integrate with Amplify Console? I'm currently developing a React SPA that doesn't use Next.js and looking for the recommended way to add SSR for SEO purposes.

**Comment by m5r** [comment]
> If I understood correctly, all the improvements that are being made to make Amplify work with Next.js will still require the final app to be hosted using Vercel, since Amplify Console doesn't support SSR. Is that right?
Vercel is one way to host your app. You can leverage the [Serverless framework](https://www.serverless.com/blog/serverless-nextjs) or just host it wherever you want to by following [their docs](https://nextjs.org/docs/deployment#nodejs-server)

**Comment by ChristopheBougere** [comment]
@ianmartorell you can upvote this issue for amplify console support ;)
https://github.com/aws-amplify/amplify-console/issues/412

**Comment by ianmartorell** [comment]
> Vercel is one way to host your app. You can leverage the [Serverless framework](https://www.serverless.com/blog/serverless-nextjs) or just host it wherever you want to by following [their docs](https://nextjs.org/docs/deployment#nodejs-server)

@m5r Thanks for the Serverless blog post link! Although I'd really like to host my app with the Amplify Console since I'm already using Amplify for everything else. And going for a node.js server just because I need SSR when I've already gone the full serverless architecture route doesn't feel great :)

> @ianmartorell you can upvote this issue for amplify console support ;)
> [aws-amplify/amplify-console#412](https://github.com/aws-amplify/amplify-console/issues/412)

@ChristopheBougere Done!

**Comment by janhesters** [comment]
Tried the `@preview` versions of `@aws-amplify/auth` and `@aws-amplify/core` out with Next.js and I get:

```
error - ./node_modules/@aws-sdk/client-cognito-identity/dist/es/runtimeConfig.browser.js
Should not import the named export 'name' (imported as 'name') from default-exporting module (only default export is available soon)
```

with both packages. When I got into the `node_modules` and into `runtimeConfig.browser.js` and instead of importing the `name` and `version` from `package.json` I hardcode them, the code works. @ericclemmons  you might need to change the way you bundle these packages 🤔 .

**Comment by ericclemmons** [comment]
@janhesters Can you share more of your code & your Next.js version?  The demo app I'm running integration tests against uses `aws-amplify@preview` as a dependency to access the `withSSRContext` helper and haven't experienced that.

**Comment by jfryerocv** [comment]
@ericclemmons will this work if someone creates a completely static app with `next export`?

**Comment by ianmartorell** [comment]
> @ericclemmons will this work if someone creates a completely static app with `next export`?

If you create a completely static app you don't need server side rendering anymore.

**Comment by ericclemmons** [comment]
@jfryerocv Yes, @ianmartorell is correct.  The majority of Amplify worked with Next.js' SSG since v3 was released earlier this year.

There is some work here to support DataStore within that environment, but otherwise the `withSSRContext` isn't needed at all for SSG.

**Comment by janhesters** [comment]
@ericclemmons I tried both, running `@preview` and the stable one and both display that behaviour. Not using `withSSRContext` because we use sagas. Instead I use `Auth` but all the code is only called on the client-side.

**Comment by ericclemmons** [comment]
@janhesters I'll check out https://github.com/aws-amplify/amplify-js/issues/6555, since that looks like the crux of the problem (and can be tested in CRA to reduce this being SSR-specific).

**Comment by janhesters** [comment]
@ericclemmons I think #6555 might have a different reason though. I'll try and create a minimal setup for https://github.com/aws-amplify/amplify-js/issues/5435#issuecomment-672741212 soon 👍

**Comment by dberhane** [comment]
It would be great to know if you have plans to support Next.js incremental static site generation and when?

**Comment by ericclemmons** [comment]
@dberhane This PR supports both SSG & SSR!  If you're using the Amplify `API` category for your data, I believe `aws-amplify@master` will already work with SSG.  For a static site (SSG), `Auth` will happen on the client & should work as usual.

**Comment by oahmaro** [comment]
is there a specific deadline for this PR? i am planing to use amplify with Next JS for my upcoming project and would like to know if this would be ready within 1 or 2 months? if so this would help me a lot as i can start implementing my business logic and UI, and add amplify once this PR is merged. 
Really appreciate the work you guys are doing here.
Thanks.

**Comment by MaazDev** [comment]
@ericclemmons Hello , is there any plan to do something similar for nuxtjs ?

**Comment by loukmane-issa** [comment]
First of all, thanks a lot @ericclemmons for all the work being done here. I have been able to get the authentication working on a page using SSR and the documentation you provided here. I have a question though, I was planning to use React Context to store if the user has been authenticated or not. This would allow me to conditionally choose layout and also make other logic for pages based on the authentication status of the user.

However, when I am trying to use the following code in the ```getInitialProps``` in my ```_app.tsx```:

```tsx
    const { Auth } = withSSRContext(context);
    let isUserAuthenticated = false;
    try {
      const user = await Auth.currentAuthenticatedUser();
      if (user) {
        isUserAuthenticated = true;
      }
    } catch (e) {
      console.error(e);
    }
```

The result always bring a non authenticated user. However, if I use the ```getInitialProps``` of the actual page, this will result in ```isUserAuthenticated``` being true. Do you know why this could happen?

Thanks.

**Comment by Nickman87** [comment]
I'm using this preview and got Authentication up and running.
However I cannot seem to get API working trough the `withSSRContext`.

When I do the following:
`const { Auth, API } = withSSRContext();`
API is null, even though the API works fine using `import { API } from "aws-amplify";`.

I'm using `"aws-amplify": "^3.0.18-preview.46",`

Is there something I can do to debug this?

Another question: when will the preview get a version update? The current version is already at 24 :)

Looking forward to get this fully up and running!

**Comment by ericclemmons** [comment]
@loukmane-issa Can you help me understand exactly what's happening here?
> The result always bring a non authenticated user. However, if I use the getInitialProps of the actual page, this will result in isUserAuthenticated being true. Do you know why this could happen?
It sounds like, _on the server_, the user isn't authenticated. But running that logic on the client works (as expected)?  I did find a bug recently with this, but haven't updated `@preview`.  Let me know if if my understanding is correct!

@Nickman87 Can you share a snippet of your code?  I was aware of this potential problem, but haven't been able to replicate.  Here's my `API` usage:

```ts
// pages/api/create-todo.tsx
import { Amplify, withSSRContext } from 'aws-amplify'
import { GraphQLResult } from '@aws-amplify/api-graphql'
import { NextApiRequest, NextApiResponse } from 'next'

import { CreateTodoMutationVariables, CreateTodoMutation } from '../../src/API'
import awsconfig from '../../src/aws-exports'
import { createTodo } from '../../src/graphql/mutations'

Amplify.configure(awsconfig)

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const { API } = withSSRContext({ req })

  try {
    const query = createTodo
    const variables: CreateTodoMutationVariables = {
      input: {
        title: `New Todo ${new Date().toLocaleString()}`,
      },
    }

    const { data } = (await API.graphql({
      query,
      variables,
    })) as GraphQLResult<CreateTodoMutation>

    return res.status(200).json({ data })
  } catch (error) {
    console.error(error)
    return res.status(500).json({ error })
  }
}
```

@Nickman87 If you're open to discussing this live, you can DM on me Twitter (@ericclemmons) or find me in our Discord: https://discord.gg/Px5YVfQ

**Comment by MontoyaAndres** [comment]
@ericclemmons is a good practice to call the method `Amplify.configure` just once in the page `_app.js` or is necessary to call it on each page?

**Comment by loukmane-issa** [comment]
@ericclemmons about:

> It sounds like, on the server, the user isn't authenticated. But running that logic on the client works (as expected)? I did find a bug recently with this, but haven't updated @preview. Let me know if if my understanding is correct!

I don't believe this is what's happening. I am not a Next.js expert, but from what I understand, The ```getInitialProps``` will either run on the server or on the client depending if it's the first load. In both cases, the authentication is working properly if I place the authentication code within the get initial props of the page itself (e.g. *pages/dashboard.tsx*). However, if I use the ```getInitialProps``` of *_app.ts*, I get no user from the ```currentAuthenticatedUser()``` method. The main reason for doing it in the _app.ts is to have the user and the authentication status available within a Context and easily access this throughout the app.

As I just started development, I have refactored my app in order to not use authentication status within a context and have the check in each page ```getInitialProps``` but I would like to go back using Context if it is possible.

**Comment by Danm72** [comment]
I'm getting no current user when using the preview

**Comment by astenmies** [comment]
Thanks so much @ericclemmons it works well in Next.js. I'm not completely done with the refactor, but if anyone needs a Next.js example, here is mine:

## Install AWS Amplify `@preview`

```sh
yarn add aws-amplify@preview

# Optional
yarn add hoist-non-react-statics
```

## Create a HOC that will be used in `_app.js`

```js
// libs/withAmplify.js
import * as React from "react";
import { Amplify, withSSRContext } from "aws-amplify";
import hoistNonReactStatics from "hoist-non-react-statics";

const AmplifyContext = React.createContext();
export const useAmplify = () => React.useContext(AmplifyContext);

export const withAmplify = (ComposedComponent) => {
	const WithAmplify = (props) => {
		const { amplifyConfig, ...rest } = props;
		const { Auth } = withSSRContext();

		React.useMemo(() => {
			Amplify.configure(amplifyConfig);
			Auth.configure(amplifyConfig);
		}, []);

		return (
			<AmplifyContext.Provider value={{ amplifyConfig }}>
				<ComposedComponent {...rest} />
			</AmplifyContext.Provider>
		);
	};

	WithAmplify.getInitialProps = async (appContext) => {
		const amplifyConfig = {
			Auth: {
				region: process.env.AWS_AMPLIFY_REGION,
				userPoolId: process.env.AWS_AMPLIFY_USER_POOL_ID,
				userPoolWebClientId: process.env.AWS_AMPLIFY_USER_POOL_WEB_CLIENT_ID,
				mandatorySignIn: true,
				authenticationFlowType: process.env.AWS_AMPLIFY_AUTHENTICATION_FLOW_TYPE,
				oauth: {
					domain: process.env.AWS_AMPLIFY_OAUTH_DOMAIN,
					scope: process.env.AWS_AMPLIFY_OAUTH_SCOPE,
					redirectSignIn: process.env.AWS_AMPLIFY_OAUTH_REDIRECT_SIGN_IN,
					redirectSignOut: process.env.AWS_AMPLIFY_OAUTH_REDIRECT_SIGN_OUT,
					responseType: process.env.AWS_AMPLIFY_OAUTH_RESPONSE_TYPE,
				},
			},
		};

		return {
			amplifyConfig,
		};
	};

	return hoistNonReactStatics(WithAmplify, ComposedComponent);
};
```

## Use withAmplify in `_app.js`

Wrap your App with `withAmplify` HOC previously created.

```js
// pages/_app.js
import * as React from "react";
import { withAmplify } from "../libs/withAmplify";

const MyApp = ({ Component, pageProps }) => {
	return <Component {...pageProps} />;
};

export default withAmplify(MyApp);
```

## Create a login page

```js
// pages/login.js
import * as React from "react";
import { withSSRContext } from "aws-amplify";
import Router from "next/router";

const Login = (props) => {
	const [state, _setState] = React.useState({
		username: "",
		password: "",
	});
	const setState = (newState) => _setState({ ...state, ...newState });

	// 👇 Your frontend code is now "SSR-aware", giving the backend access to the current session
	const { Auth } = withSSRContext();

	React.useEffect(() => {
		(async () => {
			let user, currentSession;
			try {
				user = await Auth.currentAuthenticatedUser();
			} catch (e) {
				console.error("currentAuthenticatedUser error:", e);
			}

			try {
				currentSession = await Auth.currentSession();
			} catch (e) {
				console.error("currentSession error:", e);
			}

			// if (currentSession) Router.push("/");
			console.log("login.user --", user);
			console.log("login.currentSession --", currentSession);
		})();
	}, []);

	const handleChange = (e) => {
		setState({ [e.target.id]: e.target.value });
	};

	const federatedSignIn = () => {
		Auth.federatedSignIn();
	};

	const signIn = async (e) => {
		e.preventDefault();
		try {
			await Auth.signIn(state.username, state.password);
			Router.push("/");
		} catch (err) {
			if (err.code === "InvalidParameterException" || err.code === "NotAuthorizedException") alert("Incorrect username or password.");
			else alert(err.message);
		}
	};

	return (
		<>
			<h1>Login page</h1>
			<form onSubmit={signIn}>
				<input id="username" type="text" value={state.username} onChange={handleChange} />
				<input id="password" type="password" value={state.password} onChange={handleChange} />
				<button type="submit" onClick={signIn}>
					Sign in
				</button>
			</form>
			<br />
			<button onClick={federatedSignIn}>Hosted UI</button>
		</>
	);
};

export default Login;
```

**Comment by astenmies** [comment]
While digging deeper in my implementation, I was hoping to override the path of the cookies options via [`cookieStorage`](https://github.com/aws-amplify/amplify-js/blob/main/vscode/auto_snippet_documentation.md#prefix-amplify-manual-setup-2), however, this has no effect on the [`CognitoIdentityServiceProvider`](https://github.com/aws-amplify/amplify-js/blob/main/packages/amazon-cognito-identity-js/src/CognitoUser.js#L1378) cookies. With the above example, the cookies set on `signIn()` have their path automatically set to `/login`. As a consequence, other pages can't have the logged in user on SSR.

Adding `path: "/"` here https://github.com/aws-amplify/amplify-js/commit/a579c9b66eec99e771da92e35ed179bb4f2f984d#diff-ac47a9e216a032c74ba8bf10c36119b3R106 seems to solve this.

Is there any reason not to enable the `path` of the cookies to be overriden?

Thanks in advance for your insight

**Comment by ericclemmons** [comment]
@astenmies Good catch! I've locally updated `UniversalStorage` to add `path: "/"`.

There's not a compelling reason _yet_ to allow that customization, as I'd like to get more real-world usage examples to figure out what signature `UniversalStorage` should have.  (The fact that it uses cookies behind the scenes is an implementation detail with it's own flaws)

**Comment by ericclemmons** [comment]
@dberhane Are you referring to the `revalidate` property in https://github.com/vercel/next.js/discussions/11552?

This effort is ensuring Amplify works within Next.js v9.3's SSG/SSR methods, so if you're deploying to Vercel (which takes advantage of `revalidate`), this should work as expected 👍

**Comment by ericclemmons** [comment]
@MontoyaAndres I'd recommend putting `Amplify.configure` in `_App.js`.  For `/api` routes, I don't think that's used, so you'll need to duplicate `Amplify.configure` in those files. (Or put it in a utility folder and import from there)

**Comment by MontoyaAndres** [comment]
Thank you @ericclemmons !

**Comment by Danm72** [comment]
@ericclemmons do you have any sample project you could share that shows everything working together?

I'm having a good bit of trouble and trying to piece together the configuration is a little tricky!

Really appreciate your work here!

**Comment by WojciechMatuszewski** [comment]
Hi guys, I've made a short video showcasing how to use `Auth` within `getServerSideProps`

https://www.youtube.com/watch?v=Fc5kgSq1lpw

Also here is the repo of a sample project I've made - authenticating and fetching data within `getServerSideProps`.

https://github.com/WojciechMatuszewski/next-appsync-cdk

**Comment by Danm72** [comment]
@WojciechMatuszewski  Hey! Code looks great. If you use user pool auth does it still work for you? Or what auth are you using for your schema/models?

For example:
amplify update api

> update auth
> user pool

Does it still work? I'm still getting 'no current user'

**Comment by WojciechMatuszewski** [comment]
@Danm72 I'm using user pools since I'm using Cognito (see `config.json` within `frontend` directory). I cannot speak for amplify-cli as I do not use that tool.

**Comment by dabit3** [comment]
@WojciechMatuszewski **nice video**, thanks for sharing!!

**Comment by gbouteiller** [comment]
First of all, great job for the feature!!
However, if you only use it for **Auth**, it looks like it adds a lot to the bundle size.
As `withSSRContext` is imported from `aws-amplify`, maybe it's because of a tree shaking problem?

**Comment by WojciechMatuszewski** [comment]
> @WojciechMatuszewski _nice video_, thanks for sharing!!
> 
> Edit: Just watched the whole video, amazing explanation! We will definitely be sharing this once the release is officially announced if that is ok with you.

Sure, no problem. Thanks!

**Comment by wmcons** [comment]
Thanks @ericclemmons, 

Just noticed, API.post works fine from client but server side API.post its throwing error with  'The security token included in the request is invalid.' using Auth,API from withssrcontext. Rest API is iam authenticated. But if i change the service cognito authorizer it does work. Any example with withssrcontext API.post works with iam authenticated rest api.

**Comment by Danm72** [comment]
The details in the #2381 helped me a lot to get things running

**Comment by ericclemmons** [comment]
👋 Hi everyone!

Today, I'm proud to say that we've officially released SSR support for Amplify JS!

- Twitter announcements:
	- https://twitter.com/AWSAmplify/status/1305897714002034693
	- https://twitter.com/ericclemmons/status/1305899894658314240

- Launch post:
	- https://aws.amazon.com/blogs/mobile/ssr-support-for-aws-amplify-javascript-libraries/

- Getting Started:
	- https://docs.amplify.aws/start/q/integration/next

- Documentation:
	- https://docs.amplify.aws/lib/ssr/q/platform/js

### What's changed?

There have been a few changes since `aws-amplify@preview` was published (see the resources above for more):

1. Configure your application with `Amplify.configure({ ...awsExports, ssr: true })` so that it's "SSR-aware".
1. Client-side code should stay the same as before:

	```
	import { Amplify, API } from "aws-amplify"`;

	...
	const { data } = await API.graphql({ query, variables })
	```

	Using `withSSRContext` on the client will hurt client-side bundle sizes!

1. Server-side code should use the new `withSSRContext` utility:

	```
	import { Amplify, withSSRContext } from "aws-amplify";
	...
	export async function getServerSideProps({ req }) {
 		const SSR = withSSRContext({ req });
		const { data } = await SSR.API.graphql({ query, variables });
		...
	```

	`withSSRContext` creates a new copy of `Amplify` that's scoped to a single `req`uest (including credentials!).

Check out the [Getting Started](https://docs.amplify.aws/start/q/integration/next) tutorial for a working example of this usage!

### What about deployment?


If you're site is fully static (no `getServerSideProps`), you can continue using Amplify Console!

However, if you're using `getServerSideProps` or `fallback: true`, there are other options:

	- Until [Amplify Console supports SSR](https://github.com/aws-amplify/amplify-console/issues/412) (be sure to upvote & comment!), we've used & tested [deploying with serverless framework](https://docs.amplify.aws/start/getting-started/hosting/q/integration/next).
	- Of course, several customers are [deploying with Vercel](https://nextjs.org/docs/deployment) and that should continue working, too.

### What's next?

With today's announcement, `API` (both GraphQL & REST), `Auth`, and `DataStore` are supported.

These features should also work with other frameworks (such as Nuxt.js & Gatsby), but we trust and rely on your input to make it even better.

The same goes for other features that didn't make it to release today (e.g. `Analytics`, `Storage`, etc.).  Please share SSR-specific use-cases for those categories in a [new issue](https://github.com/aws-amplify/amplify-js/issues/new/choose) on how Amplify can be better experience, and we'll be happy to help!

With that, I want to personally thank everyone on this issue for your feedback and effort.  I'm thrilled that I could play a part in bringing Next.js support to Amplify JS, and want to continue working with y'all to make it even better!

Thanks!

**Comment by MontoyaAndres** [comment]
Thank you @ericclemmons and all the team behind this awesome work! 🦾

**Comment by wmcons** [comment]
Hi @ericclemmons,

 First of all awesome work, it's working flawlessly. I am trying to look this from a different angle [Not sure whether this can be viewed this way].I wanted to login from server using api route using amplify, auth and keep the user session in server  ( works ) and send the necessary token to client side from server to access the API details. To store user session used session server like redis with express session and send the accessToken, IdToken back to client reconstruct user session. But server side when ever client token expires it can send a request to server to refresh the token and get new access/id token. Issue i am facing now because i have user session how can i use that to create Auth, API ( Cookie is not used ) as its passed in session store as user session, also in server side i have user session in memory db ( Redis for e.g ). ( I understand we can use aws amplify low level api's to construct these using id/access/refresh ) but do we have something out of the box by simply passing user session and it will construct these objects accordingly.

**Comment by omar-dulaimi** [comment]
Hello,
has this changed since you last added the docs?
`Note: Next also supports pre-rendering for dynamic server-rendered routes. At this time, Amplify does not support the hosting of dynamic server-rendered routes with Next.`

Looking for docs explaining SSR and Amplify setup.

**Comment by omar-dulaimi** [comment]
Hello,
where can I find docs that explain SSR setup using Nextjs and Amplify?

**Comment by Danm72** [comment]
@omar-dulaimi https://docs.amplify.aws/start/q/integration/next

**Comment by omar-dulaimi** [comment]
Thank you @Danm72

**Comment by flyandi** [comment]
This is awesome but the question is now, how do I use amplifyhosting to enable SSR? :-)

**Comment by ianmartorell** [comment]
> This is awesome but the question is now, how do I use amplifyhosting to enable SSR? :-)

@flyandi That's tracked in this issue: https://github.com/aws-amplify/amplify-console/issues/412

**Comment by ohlr** [comment]
One question regarding [the blog](https://docs.amplify.aws/start/getting-started/data-model/q/integration/next)

It suggests using api-keys to build the app statically. If I do that, my builds stop working as soon the key expires (after 7 days).

Is there any solution to this? @ericclemmons

**Comment by ericclemmons** [comment]
@ohlr For real-world apps, I also opt for IAM instead. (Originally, the tutorial used IAM but it has a few more prompts that are easy to miss)

When you run `amplify add api` or `amplify update api` and select IAM, but be sure to set the Create/Read/Update/Delete permissions correctly. You'll want the IAM account to read-only, and leave Cognito User Pools to the write/delete operations.

**Comment by ohlr** [comment]
@ericclemmons Thank you for the hint with the permissions. I previously did not configure iam as provider.
```
      #schema.graphql
      { allow: groups, groups: ["Admin"] }
      { allow: public, operations: [read], provider: iam }
```

**Comment by ericclemmons** [comment]
@ohlr Of course!  I should've mentioned that or shared this link:
> https://docs.amplify.aws/cli/graphql-transformer/auth#public-authorization

**Comment by hugomn** [comment]
Thanks for the great work adding SSR to amplify! I use `apollo-client` to handle our graphql queries and state, so I cannot use the `SSR.API.graphql()` call. I'd need to get the session jwt token on the server and create an apollo client with it. Is it currently possible? I tried something like below, without sucess. 

```
export const getServerSideProps: GetServerSideProps = async (context) => {
  const { Auth } = withSSRContext(context)
  const session = await Auth.currentSession()
  const token = session.idToken.jwtToken
  console.log('Apollo client token: ', token)
  return {}
}
```

Any ideas? Thanks!

**Comment by MontoyaAndres** [comment]
I'm trying to execute `Auth.signOut` but sometimes the user is logout and sometimes not. This is what I'm doing:

```
  const signOut = () => {
    Auth.signOut().then(() => {
      router.reload();
    });
  };
```

When the user executes the function signOut, the page is reloaded, then the /page/MySecretPage.tsx executes this method to validate if the user has access or not:

```
export async function getServerSideProps(context) {
  await withAuth(context);

  return {
    props: {},
  };
}
```

This is the function withAuth:

```
import { withSSRContext } from 'aws-amplify';

export const withAuth = async context => {
  try {
    const SSR = withSSRContext(context);
    const user = await SSR.Auth.currentAuthenticatedUser();

    return user;
  } catch (error) {
    context.res.writeHead(301, { Location: '/signin' });
    context.res.end();
  }
};
```

Right now I have the `Amplify.configure` on the _app.tsx file. I haven't tried on each page as the docs explain.

**Comment by dbhagen** [comment]
Wow, I'm just hitting the same issue, @MontoyaAndres! Amazing timing. Out of curiosity, are you using SSO of any kind?

**Comment by MontoyaAndres** [comment]
@dbhagen I'm not the only one, that's great :), yes, I'm using sso. I tried to combine it with SSR but it didn't work, also sometimes the method on the function getServerSideProps does not work, and throw me to login...

**Comment by dbhagen** [comment]
@MontoyaAndres so Eric asked me to fork Nader's repo for NextJS and recreate in that. I've done so here: https://github.com/dbhagen/next.js-authentication-aws

I'll admit, the example worked fine until AWS Auth/WebUI/SSO was added. If you want to be added to that repo to set up a specific page/use case, I'd be happy to add you.

**Comment by MontoyaAndres** [comment]
I've seen the code and it's implementing exactly the same I'm doing now. So, I think I'll wait until this problem with AWS Auth/WebUI/SSO is fixed :)

**Comment by ericclemmons** [comment]
@MontoyaAndres I love the detail you’ve put into this 

Can you copy/paste this info into a new issue so we can track and address it directly?

I’ve been able to validate SSR behavior with my samples locally, but there’s something amiss here...

**Comment by MontoyaAndres** [comment]
Hey @ericclemmons , yep, here it is https://github.com/aws-amplify/amplify-js/issues/7223

**Comment by devcshort** [comment]
I'm currently using Amplify's new SSR functionality in production for [WA 211](https://search.wa211.org/) and haven't run in to any issues yet. We love how quick and easy it is to set up.

I was slightly concerned that using `withSSRContext` and `SSR.Auth.currentAuthenticatedUser()` may not be secure. Is the data this retrieves fully from the server side and can't be changed on the client prior to the requests being sent? Is it secure to use the user sub directly from this method or any of the other methods as long as we're using withSSRContext?

**Comment by theelk801** [comment]
Hi, I have an [issue](https://github.com/aws-amplify/amplify-console/issues/2107) that seems to be related to this and it's been really frustrating me. Is there any way I could get help on it?
